src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServer.cs

```diff
@@ -131,8 +131,15 @@ public void Start<TContext>(IHttpApplication<TContext> application)
 
                     return;
                 }
-                else if (!hasListenOptions)
+                else if (!hasListenOptions || (_serverAddresses.PreferHostingUrls && hasServerAddresses))
```

Comment:
Not a big fan of how these if else are structured since it's somewhat convoluted. As far as I can tell, it's to reduce code duplication for translating server addresses to listen options and binding to the listen options.

I have an alternative approach added to the branch at https://github.com/aspnet/KestrelHttpServer/compare/johluo/use-hosting-urls?diff=split&expand=1&name=johluo%2Fuse-hosting-urls but there's a bit of moving things around so I'm not sure if it's worth the effort.